src/cc/libbpf.c

```diff
@@ -823,15 +817,16 @@ void *bpf_attach_kprobe(int progfd, enum bpf_probe_attach_type attach_type,
     close(kfd);
     snprintf(buf, sizeof(buf), "/sys/kernel/debug/tracing/events/%ss/%s", event_type, event_alias);
   }
-  if (bpf_attach_tracing_event(progfd, buf, reader, -1 /* PID */, pfd) < 0)
-    goto error;
-
-  return reader;
+  // If perf_event_open succeeded, bpf_attach_tracing_event will use the created
+  // Perf Event FD directly and buf would be empty and unused.
+  // Otherwise it will read the event ID from the path in buf, create the
+  // Perf Event event using that ID, and updated value of pfd.
+  if (bpf_attach_tracing_event(progfd, buf, -1 /* PID */, &pfd) == 0)
+    return pfd;
 
 error:
-  perf_reader_free(reader);
-  return NULL;
-
+  bpf_close_perf_event_fd(pfd);
```

Comment:
Do you need to check whether pfd >= 0 here? Is it possible pfd still equals to initial value -1?
The same for below a few places.

Reply:
`bpf_close_perf_event_fd` already checks the FD inside:)